src/pages/AttendancePublic.js

```diff
@@ -0,0 +1,162 @@
+    } catch (error) {
+      console.error("Error fetching sheet data:", error);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    if (gid) {
+      fetchData();
+    }
+  }, [gid]);
+
+  return (
+    <div className="mt-8 pb-8 grid grid-cols-1 justify-items-center sm:grid-cols-6">
+      {isLoading ? (
+        <span className="loading loading-dots loading-lg sm:col-span-6"></span>
+      ) : (
+        responses.length > 0 && (
+          <>
+            <div className="card w-full bg-neutral text-neutral-content sm:col-start-2 sm:col-span-4">
+              <div className="card-body items-center text-center">
+                <h1 className="card-title">{courseName}</h1>
+                <div className="card-actions justify-end">
+                  <button
+                    className="btn btn-primary btn-sm"
+                    onClick={fetchData}
+                  >
+                    REFRESH
+                  </button>
+                  <button
+                    className="btn btn-ghost btn-sm"
+                    onClick={downloadSheet}
+                  >
+                    DOWNLOAD
+                  </button>
+                </div>
+                <QuizTable responses={responses} />
+              </div>
+            </div>
+          </>
+        )
+      )}
+    </div>
+  );
+};
+
+const QuizTable = ({ responses }) => {
+  return (
+    <table className="table sm:col-span-6">
+      <thead>
+        <tr>
+          {responses[0]
+            .filter((header) =>
+              ["Timestamp", "Email Address", "Score"].includes(header)
+            )
+            .map((header, index) => (
+              <th key={index}>{header}</th>
+            ))}
+        </tr>
+      </thead>
+      <tbody>
+        {responses.slice(1).map((row, rowIndex) => {
+          const [timestamp, emailAddress, score] = row;
+          return (
+            <tr key={rowIndex}>
+              <td key={0}>{timestamp}</td>
+              <td key={1}>{emailAddress}</td>
+              <td key={2}>{score}</td>
+            </tr>
+          );
+        })}
+      </tbody>
+    </table>
+  );
+};
+
+export const AttendancePublic = () => {
+  const DB_COURSE_KEY = "courses";
+  const coursesRef = ref(db, DB_COURSE_KEY);
+  const [courseGidMap, setCourseGidMap] = useState(new Map());
+  const [courseName, setCourseName] = useState("Select Course");
+  const [gid, setGid] = useState("");
+  const [courseOptions, setCourseOptions] = useState([]);
+
+  useEffect(() => {
+    onChildAdded(coursesRef, (data) => {
+      const { courseTitle, gid } = data.val();
+      courseGidMap.set(courseTitle, gid);
+      setCourseGidMap((prevMap) => prevMap.set(courseTitle, gid));
+      const courseTitles = Array.from(courseGidMap.keys());
+      setCourseOptions(
+        courseTitles.map((title) => (
+          <option key={title} value={title}>
+            {title}
+          </option>
+        ))
+      );
+    });
```

Comment:
functionalities like these would benefit from being wrapped in a function, to give it a name. That would explain what is going on here